lib/python/picongpu/utils/picongpu_memory_calculator.py

```diff
@@ -0,0 +1,363 @@
+#!/usr/bin/env python
+
+"""
+This file is part of PIConGPU.
+
+It is supposed to give an estimate for the memory requirement of a PIConGPU
+simulation per accelerator.
+
+Copyright 2018 PIConGPU contributors
+Authors: Marco Garten
+License: GPLv3+
+"""
+
+import numpy as np
+
+
+class PicongpuMemoryCalculator:
+    """
+    Memory requirement calculation tool for PIConGPU
+
+    Contains calculation for fields, particles, random number generator
+    and the calorimeter plugin. In-situ methods other than the calorimeter
+    so far use up negligible amounts of memory on the accelerator.
+    """
+
+    def __init__(
+            self,
+            n_x,
+            n_y,
+            n_z,
+            precision_bits=32
+    ):
+        """
+        Class constructor
+
+        :param n_x: int
+            number of cells in x direction (per accelerator)
+        :param n_y: int
+            number of cells in y direction (per accelerator)
+        :param n_z: int
+            number of cells in z direction (per accelerator)
+        :param precision_bits:
+            floating point precision for PIConGPU run
+        """
+        # local accelerator domain size
+        self.n_x = n_x
+        self.n_y = n_y
+        self.n_z = n_z
+
+        self.precision_bits = precision_bits
+
+        if self.precision_bits == 32:
+            # value size in bytes
+            self.value_size = np.float32().itemsize
+        elif self.precision_bits == 64:
+            # value size in bytes
+            self.value_size = np.float64().itemsize
+
+    def mem_req_by_fields(
+            self,
+            n_x=None,
+            n_y=None,
+            n_z=None,
+            field_tmp_slots=1,
+            particle_shape_order=2,
+            sim_dim=3
+    ):
+        """
+        Memory reserved for fields on each accelerator
+
+        Parameters
+        ----------
+
+        n_x : int
+            number of cells in x direction (per accelerator)
+        n_y : int
+            number of cells in y direction (per accelerator)
+        n_z : int
+            number of cells in z direction (per accelerator)
+        field_tmp_slots : int
+            number of slots for temporary fields
+            (see PIConGPU ``memory.param`` : ``fieldTmpNumSlots``)
+        particle_shape_order : int
+            numerical order of the particle shape (see PIConGPU
+            ``species.param``: e.g. ``particles::shapes::PCS : 3rd order``)
+        sim_dim : int
+            simulation dimension (available for PIConGPU: 2 and 3)
+
+        Returns
+        -------
+
+        req_mem : int
+            required memory {unit: bytes} per accelerator
+        """
+        if n_x is None:
+            n_x = self.n_x
+        if n_y is None:
+            n_y = self.n_y
+        if n_z is None:
+            n_z = self.n_z
```

Comment:
@codingS3b Is there a better solution to this (using self values as default input arguments)?

Reply:
It's not obvious to me what is the point of having these `n_x, n_y, n_z` arguments here at all? They are all mandatory for the constructor. Or, if they are permitted to be `None` in the constructor, `n_x`, etc. could be still `None` here after these checks (in case the original value is also `None`).

Reply:
the idea was to have them be optional if you constructed the class and used the constructor but if you wanted to just know how much memory your fields use up you can just use the function

anyhow - not too much work went into the structure of this thing so suggestions are welcome =)

Reply:
After checking [stackoverflow](https://stackoverflow.com/questions/15189245/assigning-class-variable-as-default-value-to-class-method-argument) it seems that there is no better solution than the current implementation.

Reply:
Maybe you could also move the function outside of the class and within the class only create a wrapper where the member attributes are forwarded to the external function?

Reply:
@n01r did you see that comment? Would that help?

Reply:
Jup, saw and [noted](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2806#issuecomment-439960845) it :)
Could be nice to do it that way - I'll try .

Reply:
Hmm ... I played around a bit with some solutions but they don't make me happy.
I will leave that up for later.

I updated the documentation and the example and will leave it up for review after my next push.